### PR TITLE
-Updated reicast.sh installation script to pull from my latest branch…

### PR DESCRIPTION
**Pull request by free5ty1e**
…, which has been rebased from the latest upstream master and has additional improvements and features: (see post here for details: http://blog.petrockblock.com/forums/topic/configuring-controllers-in-reicast/page/2/#post-99514 )

* Launch script now runs emulator and the fallback file allocation for rpi2 from tmpfs in memory instead of performing lengthy SD writes mid-emulation (which was causing long freezes in gameplay and general overall slowdown)
* Controller mapping via emu.cfg is now a reality, individually for different controllers for different players
* Basic multiplayer is now implemented!
* 8 VMUs are now connected!
* VMUs and emu.cfg and BIOS files now all reside in home folder /home/pi/.reicast and .reicast/data/ for the BIOS files
* More improvements but I forget ATM

### CODE REVIEW

**Comment by joolswills** [comment]
@gizmo98 Please can I get your feedback on this - as the repository has changed from yours etc.

File: scriptmodules/emulators/reicast.sh [hunk 1]
```diff
@@ -40,13 +40,25 @@ function install_reicast() {
...
+sudo mount -o size=450M -t tmpfs none tmpfs/
+sudo cp -v * tmpfs/
+cd tmpfs
+sudo aoss ./reicast.elf -config config:homedir=/home/pi -config config:image="$1"
+cd ..
+echo Ensuring any freshly-created VMUs are owned by pi and not root...
+sudo chown -R pi:pi .reicast
+echo Freeing up memory...
+sudo umount /opt/retropie/emulators/reicast/tmpfs
+sudo rm -rf tmpfs
```
Comment (joolswills) [inline scriptmodules/emulators/reicast.sh]: Does this need to be absolutely pathed ? if so, you can use `$rootdir` but the mount didn't so the umount probably doesn't too if in the right folder.

**Comment by joolswills** [comment]
Apologies for messing up my comments (some on your branch, some on the diff). I didn't notice :/

**Comment by free5ty1e** [comment]
Ah, thanks for the info!  I wasn't sure what to use for the home dir instead of hardcoding, that's exactly what I needed... will update PR soon!

The tmpfs thing works great, and is cleaned up after emulator launch.  Unfortunately, Reicast falls back on file allocation for rpi2 since shm apparently does not work (at least, the way it's currently coded).. so the tmpfs thing allows this file allocation to take place in memory and actually makes the emulator playable.

It's a goal to figure out how to do this all within the main.cpp code of reicast, but for now this enables Dreamcast fun :D

**Comment by joolswills** [comment]
Did you try `sudo mount -o remount,exec /dev/shm` 

see:
https://github.com/reicast/reicast-emulator/issues/601#issuecomment-74723710

**Comment by joolswills** [comment]
Also there may be a possibility to modify the mprotect call in the code if that's what it is failing on.

**Comment by free5ty1e** [comment]
Oh hey, that's great -- I'll try that out and update the PR again :D 

Thanks!!

**Comment by free5ty1e** [comment]
Also, did you want these commits all squashed?

**Comment by joolswills** [comment]
re commits - squashed would be better thanks.

**Comment by joolswills** [comment]
reading the last of the thread linked above. looks like there can be a code fix to avoid the mount too - by calling http://man7.org/linux/man-pages/man2/personality.2.html checking what would be needed and if anyone else has already done it for this code.

**Comment by free5ty1e** [comment]
Even better!

**Comment by joolswills** [comment]
Something like

```
#include <sys/personality.h>
...
int pers, old_pers;
pers = personality(0xffffffff);
pers &= ~READ_IMPLIES_EXEC;
old_pers = personality(pers);
```

**Comment by free5ty1e** [comment]
I can confirm that remounting SHM with the command `sudo mount -o remount,exec /dev/shm` does NOT fix the issues that the tmpfs mount fixes.  The frequent emulator hangs are back as the SD card is accessed during gameplay.  :(

Will try the personality fix after my appointment.  Thanks for the assist!!  :D

**Comment by joolswills** [comment]
I guess we need to debug further - might be worth trying the code above anyway.

**Comment by joolswills** [comment]
Aah they fixed it upstream on a branch

https://github.com/reicast/reicast-emulator/commit/14875df52b547351d37d050fbd4e6f0154b343cb

**Comment by joolswills** [comment]
(assuming this is the same issue - which it may not be) - maybe worth diffing against that branch though.

**Comment by free5ty1e** [comment]
That commit / those changes have been merged in to this branch, that doesn't fix the issue either.  I still have to try messing with the personality but at the moment, the PR as is works great :D

(I took the original branch we were building for RetroPie and rebased that on the latest upstream master, and then based my changes on top of that.)

**Comment by free5ty1e** [comment]
I also have a pull request in with reicast upstream:
https://github.com/reicast/reicast-emulator/pull/666

**Comment by free5ty1e** [comment]
Ah, the personality change you referred to has also already been merged in to this branch.  The following is how it is handled in the source:

```
    printf("Personality: %08X\n", personality(0xFFFFFFFF));
    personality(~READ_IMPLIES_EXEC & personality(0xFFFFFFFF));
    printf("Updated personality: %08X\n", personality(0xFFFFFFFF));
```

**Comment by free5ty1e** [comment]
Squashed the commits.

**Comment by joolswills** [comment]
Thanks.

**Comment by free5ty1e** [comment]
Awesome!  Thanks again; I'm updating the Dreamcast retropie wiki nao :D

**Pull request merged**